File: shell.d/platform-freebsd/acpi.py
#!/usr/bin/env python2.7
import os
import re
from subprocess import check_output

VALUE_RX = re.compile('C$')
FLOAT_RX = re.compile('^\-?[0-9]+\.[0-9]+$')
INT_RX = re.compile('^\-?[0-9]+$')
ACPI_RX = re.compile('^hw\.acpi\.')
ACPI_THERMAL_RX = re.compile('^hw\.acpi\.thermal\.tz\d+\.')


def process_value(key, value):
    if key in [
        'field.PSV',
        'field.HOT',
        'field.CRT',
        'field.ACx',
        'temperature'
    ]:
        if isinstance(value, list):
            value = [VALUE_RX.sub('', i) for i in value]
        else:
            value = VALUE_RX.sub('', str(value))

    elif key in [
        'passive_cooling',
        'reset_video',
        'handle_reboot',
        'disable_on_reboot',
        'verbose',
        's4bios'
    ]:
        if value == '1':
            value = 'true'
        else:
            value = 'false'

    if value in ['-1', 'NONE']:
        return None
    else:
        return value


def get_type(value):
    if value == 'true' or value == 'false':
        return 'bool'
    elif FLOAT_RX.match(value):
        return 'float'
    elif INT_RX.match(value):
        return 'int'
    else:
        return 'str'


with open(os.devnull, 'w') as devnull:
    acpi_output = [
        i.strip() for i in check_output([
            'sysctl', '-a'
        ], stderr=devnull).split("\n") if ACPI_RX.match(i)
    ]

if len(acpi_output):
    zone_count = [i for i in acpi_output if ACPI_THERMAL_RX.match(i)]
    zone_count = [i.split('.')[3] for i in zone_count]
    zone_count = [i for i in zone_count if i is not None]
    zone_count = [i for i in zone_count if i.startswith('tz')]
    zone_count = sorted(zone_count)
    zone_count = list(set(zone_count))
    zone_count = [int(i.replace('tz', '')) for i in zone_count]
    zone_count = zone_count[-1]

    for tz in range(zone_count + 1):
        ACPI_TZ_RX = re.compile('^hw\.acpi\.thermal\.tz{0}\.'.format(tz))
        tz_output = [i for i in acpi_output if ACPI_TZ_RX.match(i)]

        if not len(tz_output):
            break

        out = []

        for i in tz_output:
            key, value = i.split(': ', 1)
            key = key.split('.')[-1]
            key = re.sub('^_', 'field.', key)

            if key in ['field.ACx', 'supported_sleep_state']:
                value = value.split(' ')

            value = process_value(key, value)

            if value is None:
                continue
            elif isinstance(value, list):
                i = 0

                for v in value:
                    v = process_value(key, v)

                    if v is not None:
                        out.append("acpi.thermal.zones.{0}.{1}.{2}:{3}:{4}".format(
                            tz, key, i, get_type(v), v
                        ))

                        i += 1
            else:
                out.append("acpi.thermal.zones.{0}.{1}:{2}:{3}".format(
                    tz, key, get_type(value), value
                ))

        for line in sorted(list(set(out))):
            if len(line):
                print(line)

File: shell.d/platform-freebsd/acpi.rb
#!/usr/bin/env ruby

def process_value(key, value)
    case key
    when 'field.PSV', 'field.HOT', 'field.CRT', 'field.ACx', 'temperature'
        if value.is_a?(Array)
            value = value.collect{|i| i.sub(/C$/, '') }
        else
            value = value.sub(/C$/, '')
        end
    when 'passive_cooling', 'reset_video', 'handle_reboot', 'disable_on_reboot', 'verbose', 's4bios'
        value = (value == '1' ? 'true' : 'false')
    end

    case value
    when '-1', 'NONE'
        return nil
    else
        return value
    end

end

def get_type(value)
    if value == 'true' || value == 'false'
        return 'bool'
    elsif value =~ /^\-?[0-9]+\.[0-9]+$/
        return 'float'
    elsif value =~ /^\-?[0-9]+$/
        return 'int'
    else
        return 'str'
    end
end

acpi_output = IO.popen("sysctl -a").read.split("\n").select{|i| i =~ /^hw\.acpi\./ }.map(&:strip).compact

if not acpi_output.empty?
    zone_count = acpi_output.select{|i| /^hw\.acpi\.thermal\.tz\d+\./ }.collect{|i|
        i.split('.')[3]
    }.compact.select{|i|
        i.start_with?('tz')
    }.sort.uniq.collect{|i|
        i.sub('tz', '').to_i
    }.last

    (zone_count + 1).times do |tz|
        tz_output = acpi_output.select{|i| /^hw\.acpi\.thermal\.tz#{ tz }\./ }

        if tz_output.empty?
            break
        end

        tz_output.collect{|i|
            key, value = i.split(': ', 2)
            key = key.split('.').last
            key = key.gsub(/^_/, 'field.')

            case key
            when 'field.ACx', 'supported_sleep_state'
                value = value.split(' ')
            end

            value = process_value(key, value)

            if value.nil?
                next
            elsif value.is_a?(Array)
                i = 0

                value.collect{|v|
                    v = process_value(key, v)

                    if v.nil?
                        nil
                    else
                        line = "acpi.thermal.zones.#{ tz }.#{ key }.#{ i }:#{ get_type(v) }:#{ v }"
                        i+=1
                        line
                    end
                }.compact
            else
                [ "acpi.thermal.zones.#{ tz }.#{ key }:#{ get_type(value) }:#{ value }" ]
            end
        }.compact.flatten.uniq.each do |line|
            puts line
        end
    end
end

# hw.acpi.thermal.tz1._TSP: 10
# hw.acpi.thermal.tz1._TC2: 5
# hw.acpi.thermal.tz1._TC1: 1
# hw.acpi.thermal.tz1._ACx: -1 -1 -1 -1 -1 -1 -1 -1 -1 -1
# hw.acpi.thermal.tz1._CRT: 106.0C
# hw.acpi.thermal.tz1._HOT: -1
# hw.acpi.thermal.tz1._PSV: 106.0C
# hw.acpi.thermal.tz1.thermal_flags: 0
# hw.acpi.thermal.tz1.passive_cooling: 1
# hw.acpi.thermal.tz1.active: -1
# hw.acpi.thermal.tz1.temperature: 29.8C
# hw.acpi.thermal.tz0._TSP: -1
# hw.acpi.thermal.tz0._TC2: -1
# hw.acpi.thermal.tz0._TC1: -1
# hw.acpi.thermal.tz0._ACx: 85.0C 55.0C 0.0C 0.0C 0.0C -1 -1 -1 -1 -1
# hw.acpi.thermal.tz0._CRT: 106.0C
# hw.acpi.thermal.tz0._HOT: -1
# hw.acpi.thermal.tz0._PSV: -1
# hw.acpi.thermal.tz0.thermal_flags: 0
# hw.acpi.thermal.tz0.passive_cooling: 0
# hw.acpi.thermal.tz0.active: 2
# hw.acpi.thermal.tz0.temperature: 27.8C
# hw.acpi.thermal.user_override: 0
# hw.acpi.thermal.polling_rate: 10
# hw.acpi.thermal.min_runtime: 0
# hw.acpi.cpu.cx_lowest: C1
# hw.acpi.reset_video: 0
# hw.acpi.handle_reboot: 1
# hw.acpi.disable_on_reboot: 0
# hw.acpi.verbose: 0
# hw.acpi.s4bios: 0
# hw.acpi.sleep_delay: 1
# hw.acpi.suspend_state: S3
# hw.acpi.standby_state: NONE
# hw.acpi.lid_switch_state: NONE
# hw.acpi.sleep_button_state: S3
# hw.acpi.power_button_state: S5
# hw.acpi.supported_sleep_state: S3 S4 S5


# echo "acpi.thermal."

# echo "os.family:str:freebsd"
# echo "os.distribution:str:$(uname -s)"
# echo "os.version:str:$(uname -r)"
# echo "os.description:str:$(uname -s) $(uname -r) $(uname -i)"
